File: app/src/lib/abschaltung.ts
/**
 * «Geht in 12 Min aus» - die Restzeit eines Ablaufs, lesbar gemacht.
 *
 * Ein Ablauf schaltet das Licht nach dreissig Minuten wieder aus. Der
 * Hub weiss das die ganze Zeit; im Haus wusste es niemand - man stand
 * im Kinderzimmer und riet, ob es gleich ausgeht oder erst in einer
 * halben Stunde. Trägt der Ablauf den Schalter «Restzeit anzeigen»,
 * schickt der Hub den Zeitpunkt als `off_at` im Zustand mit
 * (hub/core/abschaltung.py); heruntergezählt wird hier.
 *
 * Rein und ohne Netz: Der Zeitpunkt kommt einmal, die Anzeige rechnet
 * daraus bei jedem Takt neu - so kostet ein laufender Countdown keine
 * einzige Meldung mehr.
 */

/** Wie viele Sekunden noch - `null`, sobald es vorbei ist. */
export function restSekunden(offAt: unknown, jetzt: number): number | null {
  if (typeof offAt !== 'number' || !Number.isFinite(offAt)) return null;
  const rest = offAt * 1000 - jetzt;
  return rest > 0 ? Math.round(rest / 1000) : null;
}

/**
 * Die Restzeit in Worten (rein, testbar).
 *
 * Unter einer Minute sekundengenau, darüber in Minuten aufgerundet:
 * «noch 1 Min» ist ehrlicher als «noch 0 Min», solange überhaupt etwas
 * übrig ist. Ab einer Stunde mit Stunde und Minute - «noch 92 Min»
 * rechnet sonst jeder selbst um.
 */
export function restText(sekunden: number): string {
  if (sekunden < 60) return `${Math.max(1, Math.round(sekunden))} s`;
  const minuten = Math.ceil(sekunden / 60);
  if (minuten < 60) return `${minuten} Min`;
  const stunden = Math.floor(minuten / 60);
  const rest = minuten % 60;
  return rest === 0 ? `${stunden} Std` : `${stunden} Std ${rest} Min`;
}

/** «geht in 12 Min aus» - der ganze Satz, oder `null`. */
export function abschaltSatz(
  state: { off_at?: number | null } | null | undefined,
  jetzt: number
): string | null {
  const rest = restSekunden(state?.off_at, jetzt);
  return rest === null ? null : `geht in ${restText(rest)} aus`;
}

/** Die kürzeste laufende Restzeit einer Geräteliste (rein, testbar).
 *
 *  Für die Raumkarte: Brennen zwei Lichter mit Frist, ist die nächste
 *  die Auskunft, auf die es ankommt - «noch 3 Min» sagt mehr über den
 *  Raum als «noch 25 Min». */
export function naechsteAbschaltung(
  items: { state?: { off_at?: number | null } }[],
  jetzt: number
): number | null {
  const reste = items
    .map((item) => restSekunden(item.state?.off_at, jetzt))
    .filter((rest): rest is number => rest !== null);
  return reste.length === 0 ? null : Math.min(...reste);
}

File: hub/homepilot/core/abschaltung.py
"""Wann ein Gerät von selbst wieder ausgeht - und wer es sehen soll.

Ein Ablauf schaltet das Licht ein und nach dreissig Minuten wieder aus.
Der Hub weiss das die ganze Zeit; im Haus wusste es niemand. Man stand
im Kinderzimmer und riet: Geht es gleich aus oder erst in einer halben
Stunde? Wer nicht warten wollte, schaltete von Hand - und damit den
Ablauf gleich mit ins Leere.

Der Ablauf trägt darum einen Schalter «Restzeit anzeigen» (Automation.
countdown). Ist er an, schreibt die Ablauf-Maschine den Zeitpunkt des
geplanten Ausschaltens als ``off_at`` in den Zustand der Entität - eine
Unix-Sekunde, kein Text und kein Takt: Die App zählt selbst herunter,
und der Hub schickt nur zweimal etwas, beim Setzen und beim Löschen.

Bewusst im Zustand und nicht in einer eigenen Liste: So reist die
Auskunft denselben Weg wie alles andere (Schnappschuss, state_changed,
Zwischenspeicher der App) und steht damit überall dort, wo das Gerät
ohnehin schon steht - auf der Kachel, in der Raumkarte, im
«Lichter an»-Blatt. Eine eigene Liste hätte drei neue Wege gebraucht
und wäre an der ersten Stelle, die jemand vergisst, stumm geblieben.

Hier steht die entscheidbare Hälfte; das Setzen und Löschen wohnt in
core/automation.py.
"""

from __future__ import annotations

from typing import Any

#: Das Feld im Zustand der Entität. Unix-Sekunden, ``None`` heisst:
#: nichts geplant (und muss ausdrücklich gesetzt werden, sonst klebt der
#: alte Wert fest - siehe core/registry.update_state).
FELD = "off_at"

#: Schritte, die eine Wartezeit sind. Nach der ersten davon hört die
#: Vorschau auf: Was danach kommt, ist der nächste Abschnitt und nicht
#: mehr das, worauf gerade gewartet wird.
_WARTEN = ("delay", "wait_until", "fade")


def ziele_nach(actions: list[dict[str, Any]], index: int) -> list[str]:
    """Welche Geräte nach dieser Wartezeit ausgehen (rein, testbar).

    Gelesen wird nur bis zur nächsten Wartezeit: «an - 30 Min warten -
    aus - 10 Min warten - Nachricht» hat zwei Abschnitte, und während
    der ersten halben Stunde stimmt nur der erste.

    Bewusst eng: nur das ausdrückliche Ausschalten (``turn_off``). Was
    eine Szene oder ein «gemeinsam umschalten» am Ende tut, hängt vom
    Zustand in dreissig Minuten ab - eine Restzeit, die sich als falsch
    herausstellt, ist schlimmer als keine.
    """
    ziele: list[str] = []
    for action in actions[index + 1 :]:
        if not isinstance(action, dict):
            continue
        art = str(action.get("type") or "command")
        if art in _WARTEN:
            break
        if art != "command" or str(action.get("command") or "") != "turn_off":
            continue
        entity_id = str(action.get("entity_id") or "").strip()
        if entity_id and entity_id not in ziele:
            ziele.append(entity_id)
    return ziele


def rest(off_at: Any, jetzt: float) -> float | None:
    """Wie viele Sekunden noch (rein, testbar).

    ``None`` bei allem, was keine Zahl ist oder in der Vergangenheit
    liegt: Ein abgelaufener Zeitpunkt ist keine Restzeit, sondern ein
    Rest, den jemand zu löschen vergessen hat.
    """
    if not isinstance(off_at, (int, float)) or isinstance(off_at, bool):
        return None
    uebrig = float(off_at) - jetzt
    return uebrig if uebrig > 0 else None
